packages/slate/src/models/schema.js

```diff
@@ -88,225 +98,444 @@ class Schema extends Record(DEFAULTS) {
+        }
+      }
 
-    const list = Range.createList(array)
-    return list
+      for (const k in d) {
+        if (!(k in data)) return fail(NODE_DATA_UNKNOWN, k, d[k])
+      }
+    }
+
+    if (nodes) {
+      const children = node.nodes.toArray()
+      let index = 0
+      let start = 0
+      let n
+
+      for (let i = 0; i < nodes.length; i++) {
+        const def = nodes[i]
+        const { min = 0, max = Infinity } = def
+        n = index - start
+        start += min
+
+        while (n < min && n < max) {
+          const child = children[index]
+
+          if (!child) {
+            return fail(NODE_CHILD_REQUIRED, index)
+          }
+
+          if (def.kind != null && !def.kind.includes(child.kind)) {
+            if (n >= min) break
+            return fail(NODE_CHILD_KIND_INVALID, child, index)
+          }
+
+          if (def.type != null && !def.type.includes(child.type)) {
+            if (n >= min) break
+            return fail(NODE_CHILD_TYPE_INVALID, child, index)
+          }
+
+          n++
+          index++
+        }
+      }
+    }
   }
 
   /**
-   * Validate an `object` against the schema, returning the failing rule and
-   * value if the object is invalid, or void if it's valid.
+   * Normalize a `node` with `reason` and `rule`.
    *
-   * This method is private, because it should always be called on one of the
-   * often-changing immutable objects instead, since it will be memoized for
-   * much better performance.
-   *
-   * @param {Mixed} object
-   * @return {Object|Void}
+   * @param {Change} change
+   * @param {Node} node
+   * @param {Object} rule
+   * @param {String} reason
+   * @param {Mixed} ...args
    */
 
-  __validate(object) {
-    let value
+  normalizeNode(change, node, reason, ...args) {
+    const rule = this.getRule(node)
+    if (!rule) return
+
+    function setNodeData(data) {
+      change.setNodeByKey(node.key, { data })
+      logger.debug('Normalized by setting `data`.', { node, data })
+    }
+
+    function setNodeIsVoid(isVoid) {
+      change.setNodeByKey(node.key, { isVoid })
+      logger.debug('Normalized by setting `isVoid`.', { node, isVoid })
+    }
+
+    function removeNode() {
+      if (node.kind == 'document') return insertDefaults()
+      change.removeNodeByKey(node.key)
+      logger.debug('Normalized by removing the node entirely.', { node })
+    }
+
+    function insertDefaults() {
+      const { nodes } = rule.defaults
+      if (!nodes) return removeNode()
+      const children = Node.createList(nodes || [])
+      node.nodes.forEach(child => change.removeNodeByKey(child.key, { normalize: false }))
+      children.forEach((child, i) => change.insertNodeByKey(node.key, i, child))
+      logger.debug('Normalized by replacing with default children.', { node, children })
+    }
+
+    function insertDefaultChild(index) {
+      const { nodes } = rule.defaults
+      if (!nodes) return removeNode()
+      const def = nodes[index]
```

Comment:
I'm a bit lost here. `index` is too generic, I don't know if it refers to the index of a child node, or some index in a rule's array. It is used to retrieve the definition for the default node, and for inserting a new child as well.